tests/unit/test_req.py

```diff
@@ -460,3 +469,132 @@ def test_filter_install():
         INFO, 'foo bar')
     assert _filter_install('I made a SyntaxError') == (
         INFO, 'I made a SyntaxError')
+
+
+def test_join_line_continuations():
+    """
+    Test joining of line continuations
+    """
+
+    lines = dedent('''\
+    line 1
+    line 2:1 \\
+    line 2:2
+    line 3:1 \\
+    line 3:2 \\
+    line 3:3
+    line 4
+    ''').splitlines()
+
+    expect = [
+        'line 1',
+        'line 2:1 line 2:2',
+        'line 3:1 line 3:2 line 3:3',
+        'line 4',
+    ]
+
+    assert expect == list(join_lines(lines))
+
+
+def test_parse_editable_from_requirements():
+    lines = [
+        '--editable svn+https://foo#egg=foo',
+        '--editable=svn+https://foo#egg=foo',
+        '-e svn+https://foo#egg=foo'
+    ]
+
+    res = [parse_line(i) for i in lines]
+    assert res == [(REQUIREMENT_EDITABLE, 'svn+https://foo#egg=foo')] * 3
+
+    req = next(parse_content('fn', lines[0]))
+    assert req.name == 'foo'
+    assert req.link.url == 'svn+https://foo#egg=foo'
+
+
+@pytest.fixture
+def session():
+    return PipSession()
+
+
+@pytest.fixture
+def finder(session):
+    return PackageFinder([], [], session=session)
+
+
+def test_parse_options_from_requirements(finder):
+    assert parse_line('-i abc') == (OPTION, ('-i', 'abc'))
+    assert parse_line('-i=abc') == (OPTION, ('-i', 'abc'))
+    assert parse_line('-i  =  abc') == (OPTION, ('-i', 'abc'))
+
+    assert parse_line('--index-url abc') == (OPTION, ('--index-url', 'abc'))
+    assert parse_line('--index-url=abc') == (OPTION, ('--index-url', 'abc'))
+    assert parse_line('--index-url   =   abc') == (
+        OPTION, ('--index-url', 'abc')
+    )
+
+    with pytest.raises(RequirementsFileParseError):
+        parse_line('--allow-external')
+
+    res = parse_line('--extra-index-url 123')
+    assert res == (OPTION, ('--extra-index-url', '123'))
+
+    next(parse_content('fn', '-i abc', finder=finder), None)
+    assert finder.index_urls == ['abc']
+
+
+def test_parse_flags_from_requirements(finder):
+    assert parse_line('--no-index') == (FLAG, ('--no-index'))
+    assert parse_line('--no-use-wheel') == (FLAG, ('--no-use-wheel'))
+
+    with pytest.raises(RequirementsFileParseError):
+        parse_line('--no-use-wheel true')
+
+    next(parse_content('fn', '--no-index', finder=finder), None)
+    assert finder.index_urls == []
+
+
+def test_get_requirement_options():
+    res = parse_requirement_options('--install-option="--abc --zxc"')
+    assert res == {'install_options': ['--abc --zxc']}
+
+    res = parse_requirement_options('--global-option "--abc"')
+    assert res == {'global_options': ['--abc']}
+
+    line = (
+        'INITools==2.0 '
+        '--global-option="--one --two -3" '
```

Comment:
The key thing here is that by passing as a single argument, we avoid getting into complexities about quoting, etc. For example, Windows and Unix users could legitimately have different views on the meaning of ```--global-option="--author='Paul Moore'"``` (On Windows, single quote does not protect embedded whitespace on the command line). With the no-flattening approach, the canonical form is ```--global-option="--author=Paul Moore"```, with no quoting confusion.

> Should --global-option='--author --author-email' really be handled by pip?

I'm not sure what you mean by "handled by pip". That defines a single global option, which should be passed to the setup.py invocation as a single option. Because we use Popen to call setup.py, we can pass it as a single argument in the args list and let subprocess handle any quoting that's needed by the OS. Pip doesn't get involved, it just passes along what you supply.

With regard to the tests, the point is that we don't need to test how badly formed input is handled, and ```--install-option="--abc --zxc"``` is clearly badly formed (notwithstanding the comment above).